lib/rules/avoid-leaking-state-in-ember-objects.js

```diff
@@ -0,0 +1,78 @@
+'use strict';
+
+const ember = require('../utils/ember');
+const utils = require('../utils/utils');
+
+
+//------------------------------------------------------------------------------
+// Ember object rule - Avoid leaking state
+// (Don't use arrays or objects as default props)
+//------------------------------------------------------------------------------
+
+module.exports = {
+  meta: {
+    docs: {
+      description: 'Avoids state leakage',
+      category: 'Ember Object',
+      recommended: true
+    },
+    fixable: null, // or "code" or "whitespace"
+    schema: [{
+      type: 'array',
+      items: { type: 'string' },
+    }],
+  },
+
+  create(context) {
+    const defaultIgnoredProperties = [
+      'classNames',
+      'classNameBindings',
+      'actions',
+      'concatenatedProperties',
+      'mergedProperties',
+      'positionalParams',
+      'attributeBindings',
+      'queryParams',
+      'attrs',
+    ];
+
+    let ignoredProperties = context.options[0] || defaultIgnoredProperties;
+
+    const message = 'Only string, number, symbol, boolean, null, undefined, and function are allowed as default properties';
+
+    const report = function (node) {
+      context.report(node, message);
+    };
+
+    const isAllowed = function(property) {
+      const value = property.value;
+      return ember.isFunctionExpression(value) ||
+        utils.isLiteral(value) ||
+        (
+          utils.isIdentifier(value) &&
+          value.name === "undefined"
+        ) ||
+        (
+          utils.isCallExpression(value) &&
+          utils.isIdentifier(value.callee) &&
```

Comment:
I think we will have to allow all `CallExpression`s.  If we don't things like this would be disallowed:

```js
Foo.extend({
  foo: computed(function() {})
});

Foo.extend({
  foo: task(function*() {})
});
```

Reply:
Actually, these cases works fine even with the above check as these are functionExpression and we are checking for isFunctionExpression and which returns true, if you suggest removing this check/condition as there might be other cases which could fail, please let me know.

Reply:
seems like service injection will fail if we don't allow only CallExpressions, so removing other checks.
Foo.extend({
  foo: inject.service()
});